File: AkkaSingleFs/ExceptionHandling.fs
module ExceptionHandling

open System
open Akka.Actor

type RandomBomb() =
    inherit UntypedActor()
    
    let r = Random()
    
    override my.OnReceive message =
        match message with
        | :? int as n -> printfn $"1000 / %d{n} = %d{1000/n}"
        | _ -> ()
        my.Self.Tell(r.Next 10)
        
    override my.PreStart() =
        printfn "Starting..."
        
    override my.PostStop() =
        printfn "Stopping..."
        
    override my.PostRestart reason =
        printfn $"Restarted! Reason = %s{reason.Message}"
        my.Self.Tell "restarted!"

type ParentActor() =
    inherit UntypedActor()
    
    let r = Random()
    let mutable child = 0
    
    override my.OnReceive message =
        if "start".Equals(message) then
            child <- child + 1
            ParentActor.Context.ActorOf(Props.Create<RandomBomb>(), child.ToString())
                       .Tell("run!")
                       
    override my.SupervisorStrategy() =
        OneForOneStrategy(3, 1000,
                          fun ex ->
                              let is_head = r.Next(2) = 0
                              printfn "Coin is %s" (if is_head then "head, restart!" else "tail, stop! Press ENTER to end.")
                              if is_head then Directive.Restart else Directive.Stop
                              )
        
let Example (actor_system: ActorSystem) =
    let parent = actor_system.ActorOf(Props.Create<ParentActor>(), "parent")
    
    parent.Tell "start"
    
    printf "Wait for bomb.."
    Console.ReadLine() |> ignore
    actor_system.Terminate() |> ignore

File: AkkaSingle/ExceptionHandling.cs
using Akka.Actor;

namespace AkkaSingle
{
    sealed class ParentActor : UntypedActor
    {
        readonly Random r = new Random();
        int child;
        protected override void OnReceive(object message) {

            if ("start".Equals(message))
                Context.ActorOf(Props.Create(() => new RandomBomb()), (++child).ToString())
                       .Tell("run!");
        }

        protected override SupervisorStrategy SupervisorStrategy() =>
            new OneForOneStrategy(maxNrOfRetries: 3, withinTimeMilliseconds: 1000,
              ex => {
                  var isHead = r.Next(2) == 0;
                  Console.WriteLine("Coin is {0}", isHead? "head, restart!" : "tail, stop! Press ENTER to end.");
                  return isHead? Directive.Restart : Directive.Stop;
              });
    }

    sealed class RandomBomb : UntypedActor
    {
        readonly Random r = new Random();
        protected override void OnReceive(object message) {
            if (message is int n)
                Console.WriteLine("1000 / {0} = {1}", n, 1000 / n);

            Self.Tell(r.Next(10));
        }

        protected override void PreStart() {
            Console.WriteLine("Starting...");
        }

        protected override void PostStop() {
            Console.WriteLine("Stopping...");
        }

        protected override void PostRestart(Exception reason) {
            Console.WriteLine($"Restarted! Reason = {reason.Message}");
            Self.Tell("restarted!");
        }
    }

    /// <summary>
    /// A simplified example of error handling
    /// ref: https://github.com/petabridge/akka-bootcamp/blob/master/src/Unit-1/lesson4/README.md
    /// </summary>
    static class ExceptionHandling
    {
        public static void Example(ActorSystem actorSystem) {
            var parent = actorSystem.ActorOf(Props.Create(() => new ParentActor()), "parent");

            parent.Tell("start");

            Console.Write("Wait for bomb..");
            Console.ReadLine();
            actorSystem.Terminate();
        }
    }
}